Sources/NIO/MultiThreadedEventLoopGroup.swift

```diff
@@ -0,0 +1,345 @@
+    }
+
+    /// Returns the next `EventLoop` from this `MultiThreadedEventLoopGroup`.
+    ///
+    /// `MultiThreadedEventLoopGroup` uses _round robin_ across all its `EventLoop`s to select the next one.
+    ///
+    /// - returns: The next `EventLoop` to use.
+    public func next() -> EventLoop {
+        return eventLoops[abs(index.add(1) % eventLoops.count)]
+    }
+
+    /// Shut this `MultiThreadedEventLoopGroup` down which causes the `EventLoop`s and their associated threads to be
+    /// shut down and release their resources.
+    ///
+    /// Even though calling `shutdownGracefully` more than once should be avoided, it is safe to do so and execution
+    /// of the `handler` is guaranteed.
+    ///
+    /// - parameters:
+    ///    - queue: The `DispatchQueue` to run `handler` on when the shutdown operation completes.
+    ///    - handler: The handler which is called after the shutdown operation completes. The parameter will be `nil`
+    ///               on success and contain the `Error` otherwise.
+    public func shutdownGracefully(queue: DispatchQueue, _ handler: @escaping (Error?) -> Void) {
+        // This method cannot perform its final cleanup using EventLoopFutures, because it requires that all
+        // our event loops still be alive, and they may not be. Instead, we use Dispatch to manage
+        // our shutdown signaling, and then do our cleanup once the DispatchQueue is empty.
+        let g = DispatchGroup()
+        let q = DispatchQueue(label: "nio.shutdownGracefullyQueue", target: queue)
+        let wasRunning: Bool = self.shutdownLock.withLock {
+            // We need to check the current `runState` and react accordingly.
+            switch self.runState {
+            case .running:
+                // If we are still running, we set the `runState` to `closing`,
+                // so that potential future invocations know, that the shutdown
+                // has already been initiaited.
+                self.runState = .closing([])
+                return true
+            case .closing(var callbacks):
+                // If we are currently closing, we need to register the `handler`
+                // for invocation after the shutdown is completed.
+                callbacks.append((q, handler))
+                self.runState = .closing(callbacks)
+                return false
+            case .closed(let error):
+                // If we are already closed, we can directly dispatch the `handler`
+                q.async {
+                    handler(error)
+                }
+                return false
+            }
+        }
+
+        // If the `runState` was not `running` when `shutdownGracefully` was called,
+        // the shutdown has already been initiated and we have to return here.
+        guard wasRunning else {
+            return
+        }
+
+        var result: Result<Void, Error> = .success(())
+
+        for loop in self.eventLoops {
+            g.enter()
```

Comment:
I know this is old code, but it feels odd that `g` is declared half a mile above

Reply:
Yeah, it's not great. For now I'm proposing we leave it.